Source/UI/NewUI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift

```diff
@@ -37,34 +35,44 @@ final class DefaultBillingFormViewModel: BillingFormViewModel {
         self.editDelegate = view
         return view
     }
-    
-    func getCell( row: Int,  delegate: BillingFormTextFieldCellDelegate?) -> UITableViewCell {
-        let currentStyle = style.cells[row]
-        guard style.cells.count > row,
-              var cellStyle = currentStyle.style
+
+    func getCell(row: Int,  delegate: CKOCellTextFieldDelegate?) -> UITableViewCell {
+        guard style.cells.count > row, var viewStyle = style.cells[row].style
         else { return UITableViewCell() }
-        
-        if cellStyle.isOptional {
-            textValueOfCellType[currentStyle.index, default: ""] += ""
-            errorFlagOfCellType[currentStyle.index] = false
+
+        let cellStyle = style.cells[row]
+        updateStyle(with: &viewStyle, cellStyle: cellStyle, tag: row)
+        return getCell(with: &viewStyle, cellStyle: cellStyle, tag: row)
+    }
+
+    private func updateStyle(with viewStyle: inout CKOCellTextFieldStyle, cellStyle: BillingFormCell, tag: Int) {
+        if viewStyle.isOptional {
+            textValueOfCellType[cellStyle.index, default: ""] += ""
+            errorFlagOfCellType[cellStyle.index] = false
         }
-        
-        let hash = currentStyle.index
-        updateErrorView(with: &cellStyle, hashValue: hash)
-        updateTextField(with: &cellStyle, hashValue: hash)
-        
-        let cell = CKOTextFieldCell()
-        cell.delegate = delegate
-        cell.update(cellStyle: currentStyle, style: cellStyle, tag: row)
+
+        let hash = cellStyle.index
+        updateErrorView(with: &viewStyle, hashValue: hash)
+        updateTextField(with: &viewStyle, hashValue: hash)
```

Comment:
I don't understand this behaviour at all and why you ever even need it.

Normally for a tableView you have some form of dataSource which will provide you with the numberOfRows as well as cellForIndexPath. I think this hash is just a hack to cover the lack of correct tableView implementation.

Reply:
No it's not a cover

Reply:
But do you think that correctly implementing dataSource for the tableView would still require a `hash`?